src/Inciter/Diagnostics.C

```diff
@@ -0,0 +1,141 @@
+}
+
+bool
+Diagnostics::compute( Discretization& d, const tk::Fields& u )
+// *****************************************************************************
+//  Compute diagnostics, e.g., residuals, norms of errors, etc.
+//! \param[in] d Discretization proxy to read from
+//! \param[in] u Current solution vector
+//! \return True if diagnostics have been computed
+// *****************************************************************************
+{
+  // Optionally collect diagnostics and send for aggregation across all workers
+  if ( !((d.It()+1) % g_inputdeck.get< tag::interval, tag::diag >()) ) {
+
+    // Collect analytical solutions (if available) from all PDEs. Note that
+    // calling the polymorphic PDE::initialize() is assumed to evaluate the
+    // analytical solution for a PDE. For those PDE problems that have
+    // analytical solutions, this is the same as used for setting the initial
+    // conditions, since if the analytical solution is available for a problem,
+    // it is (so far anyway) always initialized to its analytical solution and
+    // that is done by calling PDE::initialize(). If the analytical solution is
+    // a function of time, that is already incorporated in setting initial
+    // conditions. For those PDEs where the analytical solution is not
+    // available, initialize() returns the initial conditions (obviously), and
+    // thus the "error", defined between this "analytical" and numerical
+    // solution will be a measure of the "distance" between the initial
+    // condition and the current numerical solution. This is not necessarily
+    // useful, but simplies the logic because all PDEs can be treated as being
+    // able to compute an error based on some "analytical" solution, which is
+    // really the initial condition.
+    auto a = u;
+    for (const auto& eq : g_pdes)
+      eq.initialize( d.Coord(), a, d.T()+d.Dt(), d.Gid() );
+
+    // Prepare for computing diagnostics. Diagnostics are defined as some norm,
+    // .e.g., L2 norm, of a quantity, computed in mesh nodes, A, as ||A||_2 =
+    // sqrt[ sum_i(A_i)^2 V_i ], where the sum is taken over all mesh nodes and
+    // V_i is the nodal volume. We send multiple sets of quantities to the host
+    // for aggregation across the whole mesh: (1) the numerical solutions of all
+    // components of all PDEs, and their error, defined as A_i = (a_i - n_i),
+    // where a_i and n_i are the analytical and numerical solutions at node i,
+    // respectively. The final aggregated solution will end up in
+    // Transporter::diagnostics().
+
+    // Diagnostics vector (of vectors) during aggregation:
+    // 0: L2-norm of all scalar components of the numerical solution
+    // 1: L2-norm of all scalar components of the numerical-analytic solution
+    // 2: Linf-norm of all scalar components of the numerical-analytic solution
+    std::vector< std::vector< tk::real > >
+      diag( 6, std::vector< tk::real >( u.nprop(), 0.0 ) );
```

Comment:
Again, perhaps good to get rid of the magic number (6)

Reply:
Sure, see d5dee471 .

Reply:
Sure, see d5dee471.